x-pack/plugins/task_manager/server/task_running/errors.ts

```diff
@@ -32,45 +41,55 @@ export class EphemeralTaskRejectedDueToCapacityError extends Error {
   }
 }
 
-function isTaskManagerError(error: unknown): error is DecoratedError {
-  return Boolean(error && (error as DecoratedError)[code]);
+function isTaskManagerError(error: unknown): error is TaskRunError {
+  return Boolean(error && (error as TaskRunError)[code]);
 }
 
-export function isUnrecoverableError(error: Error | DecoratedError) {
+export function isUnrecoverableError(error: Error | TaskRunError) {
   return isTaskManagerError(error) && error[code] === CODE_UNRECOVERABLE;
 }
 
-export function throwUnrecoverableError(error: Error) {
-  (error as DecoratedError)[code] = CODE_UNRECOVERABLE;
+export function throwUnrecoverableError(error: Error, errorSource = TaskErrorSource.FRAMEWORK) {
+  (error as TaskRunError)[code] = CODE_UNRECOVERABLE;
+  (error as TaskRunError)[source] = errorSource;
   throw error;
 }
 
-export function isRetryableError(error: Error | DecoratedError) {
+export function isRetryableError(error: Error | TaskRunError) {
   if (isTaskManagerError(error) && error[code] === CODE_RETRYABLE) {
     return error[retry];
   }
   return null;
 }
 
-export function createRetryableError(error: Error, shouldRetry: Date | boolean): DecoratedError {
-  (error as DecoratedError)[code] = CODE_RETRYABLE;
-  (error as DecoratedError)[retry] = shouldRetry;
-  return error;
-}
-
-export function throwRetryableError(error: Error, shouldRetry: Date | boolean) {
-  throw createRetryableError(error, shouldRetry);
+export function throwRetryableError(
+  error: Error,
+  shouldRetry: Date | boolean,
+  errorSource = TaskErrorSource.FRAMEWORK
+) {
+  (error as TaskRunError)[code] = CODE_RETRYABLE;
```

Comment:
I'd rather we do a check the error type and cast the error once, but we can clean it up before finalizing ...